README.md
# ting_recommender

Add the panel Ting->Ting Recommender to the object display in /admin/structure/pages/edit/ting_object

Use #ting-recommender as placement container in the bibspire admin interface.

## The widget:
Because on EasyOPAC instances we need some customizations to be done, was decided
to use the local variant of Bibspire widget (https://cdn.bibspire.dk/ddbcms.js) so
the source code of the widget was implemented into present module (original repo: https://gitlab.com/bibspire/bibspire-widget).

In our case file `js/widget/recommender.js` is being attached to the Ting Recommender panel pane and make asynchronous request
to external service for fetching recommendation items and render those.

### Usage:
The widget script is automatically added to the needed pages with Ting Recommender panel pane.

### Development:
1. Before any actions run `npm install` from the `src/` directory.
2. Do needed changes in code files.
3. Run `npm run deploy` to have `recommender.js` file compiled. The `recommender.js` file will be automatically placed
in `js/widget` directory from where this is served.


CHANGELOG.md
## Change Log

### 7.x-6.2.1 (10/06-2021)
- [#12](https://github.com/easySuite/ting_recommender/pull/12) EASYOPAC-1281 - Clean module. (@bozeak)
- [#7](https://github.com/easySuite/ting_recommender/pull/7) NO-TASK - Fix non-persistent JS errors. (@bozeak)

### 7.x-6.1.0 (14/09-2020)
- [#6](https://github.com/easySuite/ting_recommender/pull/6) EASYOPAC-1248 - Make JS compatible with IE. (@bozeak)

### 7.x-6.0.2 (15/06-2020)
- [#5](https://github.com/easySuite/ting_recommender/pull/5) NO-TASK - Translate ting recommender block heading on user profile. (@bozeak)
- [#4](https://github.com/easySuite/ting_recommender/pull/4) NO-TASK - Make heading text more consistent. (@bozeak)
- [#3](https://github.com/easySuite/ting_recommender/pull/3) NO-TASK - Remove non-existing panes. (@bozeak)

### 7.x-6.0.1 (15/04-2020)
- [#2](https://github.com/easySuite/ting_recommender/pull/2) SUP-1811 - Translation of “Inspiration” from ting_recommender module. (@bozeak)

### 7.x-5.0.0 (25/11-2019)
- [#1](https://github.com/easySuite/ting_recommender/pull/1) EASYOPAC-1195 - Implement recommender through features override. (@bozeak)

js/src/index.js
/*! DDB-CMS recommender widget, copyright Veduz ApS */
import "core-js/stable";
import "regenerator-runtime/runtime";
import React from "react";
import ReactDOM from "react-dom";
import RecommendationRows from "./RecommendationRows";
import RecommendationCols from "./RecommendationCols";
import RecommendationLines from "./RecommendationLines";
import axios from "axios";
const _ = {
  uniq: require("lodash/uniq")
};
//const sleep = ms => new Promise(resolve => setTimeout(resolve, ms));

const urlObjectMatch = location.href.match(/\/ting\/object\/([^?#.]*)/);
const urlCollectionMatch = location.href.match(/\/ting\/collection\/([^?#.]*)/);
let started = false;
let relatedPromise;
if (urlObjectMatch) {
  relatedPromise = axios.get(
    `https://api.bibspire.dk/v1/recommend?pid=${urlObjectMatch[1]}&site=${
      location.host
    }&token=l5ICyj27g8QeQcus`
  );
}

if (urlCollectionMatch) {
  relatedPromise = axios.get(
    `https://api.bibspire.dk/v1/recommend?pid=${urlCollectionMatch[1]}&site=${
      location.host
    }&token=l5ICyj27g8QeQcus`
  );
}

const urlStatusLoanMatch = location.href.match(/\/user\/me\/status/);

const RelatedRows = ({ related, src }) => (
  <div>
    <h2 style={{ marginTop: 72 }} className="sub-menu-title">
      {Drupal.t("Prøv også:")}
    </h2>
    {related.slice(0, 10).map(({ name, image, creator, hasPart, wid }) => (
      <a
        key={wid}
        style={{ marginBottom: 20, display: "block" }}
        href={
          "/ting/object/" + hasPart[0].pid + "?u_navigatedby=bibspire_" + src
        }
      >
        <img
          style={{
            verticalAlign: "top",
            display: "inline",
            opacity: 0.7,
            marginRight: 10
          }}
          src={image}
          width={80}
        />
        <span
          style={{
            verticalAlign: "top",
            display: "inline-block",
            width: 180
          }}
        >
          <h3
            className="item-title has-message"
            style={{ margin: 0, padding: 0 }}
          >
            {name}
          </h3>
          <div className="item-creators">{creator && creator[0]}</div>
        </span>
      </a>
    ))}
  </div>
);

async function attach() {
  if (started) {
    return;
  }
  if (urlObjectMatch) {
    started = true;

    const pid = urlObjectMatch[1];
    let related = (await relatedPromise).data;
    const config = (related.config && related.config.tingObject) || {};

    const parent =
      document.querySelector("#ting-recommender") ||
      document.querySelector(".ting-object");
    related = related && related.related;
    if (related && parent) {
      const elem = document.createElement("div");
      parent.appendChild(elem);
      ReactDOM.render(
        <RecommendationCols related={related} config={config} />,
        elem
      );
    }
  }

  if (urlCollectionMatch) {
    started = true;

    const pid = urlCollectionMatch[1];
    let related = (await relatedPromise).data;
    const config = (related.config && related.config.tingObject) || {};

    related = related && related.related;
    if (related && related.length) {
      if (window.innerWidth <= 950) {
        const parent = document.querySelector(".primary-content");
        const elem = document.createElement("div");
        parent.appendChild(elem);
        ReactDOM.render(
          <div style={{ clear: "both", paddingTop: 120 }}>
            <RecommendationCols related={related} config={config} />
          </div>,
          elem
        );
      } else {
        const parent = document.querySelector("#ting-recommender");
        const elem = document.createElement("div");
        parent.appendChild(elem);
        ReactDOM.render(
          <div style={{ marginTop: 150 }}>
            <RelatedRows related={related} src="collection" />
          </div>,
          elem
        );
      }
    }
  }

  if (urlStatusLoanMatch) {
    started = true;
    let loans;
    //for (let i = 0; i < 20; ++i) {
    loans = _.uniq(
      Array.from(document.querySelectorAll(".primary-content a"))
        .map(o => o.href)
        .filter(s => s.match(/ting\/object\//))
        .map(o =>
          o.replace(/.*\/ting\/object\/([^?#]*).*/, (_, _1) =>
            decodeURIComponent(_1)
          )
        )
    );
    /*
      if (loans.length) {
        break;
      }
      await sleep(200);
    }
    */
    relatedPromise = axios.get(
      `https://api.bibspire.dk/v1/recommend?${loans
        .map(s => "pid=" + s)
        .join("&")}&site=${location.host}&token=l5ICyj27g8QeQcus`
    );
    window.p = relatedPromise;
    let related = (await relatedPromise).data;
    const config = (related.config && related.config.tingObject) || {};
    related = related && related.related;
    if (related && related.length) {
      if (window.innerWidth <= 950) {
        const parent = document.querySelector(".primary-content");
        const elem = document.createElement("div");
        parent.appendChild(elem);
        ReactDOM.render(
          <div style={{ clear: "both", paddingTop: 120 }}>
            <RecommendationCols related={related} config={config} />
          </div>,
          elem
        );
      } else {
        const parent = document.querySelector(".secondary-content");
        const elem = document.createElement("div");
        parent.appendChild(elem);
        ReactDOM.render(
          <RelatedRows related={related} src="userstatus" />,
          elem
        );
      }
    }
  }

  if (
    location.href === "http://localhost:8080/" ||
    location.hash === "#bibspire-layout-demo"
  ) {
    started = true;
    const pid = "870970-basis:54970021";
    relatedPromise = axios.get(
      `https://api.bibspire.dk/v1/recommend?pid=${pid}&site=${location.host}&token=devtoken`
    );
    const related = (await relatedPromise).data;
    const elem = document.createElement("div");
    document.body.appendChild(elem);
    ReactDOM.render(
      <RecommendationCols
        related={related.related}
        config={related.config.tingObject}
      />,
      elem
    );
  }
}

if (
  document.readyState === "complete" ||
  document.readyState === "interactive"
) {
  attach();
} else {
  document.addEventListener("DOMContentLoaded", attach);
}


js/ting_recommender.js
/**
 * @file
 * JS logic used in ting_recommender.
 */

"use strict";

window.addEventListener('load', function () {
  var sidebar = document.querySelector('.secondary-content');

  var tingObject = location.href.match(/\/ting\/object\/([^?#.]*)/);
  var collection = location.href.match(/\/ting\/collection\/([^?#.]*)/);
  var user = location.href.match(/\/user\/me\/status/);

  if (tingObject) {
    var container = document.getElementById('ting-recommender');
    if (container.childElementCount !== 0) {
      var headingGrid = container.getElementsByTagName('h2').item(0);
      headingGrid.textContent = Drupal.t('Inspiration');
      container.removeAttribute('style');
    }
  }

  if ((user || collection) && sidebar !== null) {
    var options = {
      childList: true,
      subtree: true
    };

    var mCallback = function (mutations) {
      for (var index = 0; index < mutations.length; index++) {
        var mutation = mutations[index];
        if (mutation.type === 'childList') {
          if (mutation.target.localName === 'div') {
            var blockWrapper = mutation.addedNodes;
            var titleWrapper = blockWrapper[0].firstChild;
            titleWrapper.textContent = Drupal.t('Prøv også:');
          }
        }
      }
    };

    var observer = new MutationObserver(mCallback);

    observer.observe(sidebar, options);

    observer.takeRecords();
  }
});


js/src/RecommenderWidget.js
/*! DDB-CMS recommender widget, copyright Veduz ApS */
import React from "react";
import ReactDOM from "react-dom";
import Mustache from "mustache";

class RecommenderWidget extends React.Component {
  constructor(props) {
    super(props);
    this.ref = React.createRef();
    this.state = {};
  }

  componentDidMount() {
    if (this.ref.current) {
      this.setState({ totalWidth: this.ref.current.offsetWidth });
    }
  }

  render() {
    setTimeout(() => {
      if (this.ref.current) {
        this.setState({ totalWidth: this.ref.current.offsetWidth });
      }
    }, 0);
    let { related, config } = this.props;
    if (config.widgetType === "custom") {
      try {
        return (
          <div
            dangerouslySetInnerHTML={{
              __html: Mustache.render(config.customHtml, {
                related: related
                  .slice(0, config.customLimit || 6)
                  .map(({ hasPart, image, name, creator, datePublished }) => ({
                    title: name,
                    image,
                    link: `/ting/object/${hasPart[0].pid}?u_navigatedBy=bibspire`,
                    creator: creator,
                    year: hasPart[0].datePublished
                  }))
              })
            }}
          />
        );
      } catch (e) {
        return "Fejl i konfiguration af egen tilpasning af anbefalinger:" + e;
      }
    }
    if (!related || related.length === 0) {
      return null;
    }
    const totalWidth = this.state.totalWidth || 100;
    const rows = config.rows || 1;
    const horizontalSpacing = 24;
    const width = +config.coverWidth || 158;
    const coverRatio = 1.398734;
    const coverHeight = width * coverRatio;
    const fontSize = 14;
    const lineHeight = 20;
    const coverTextSpacing = 8;
    const verticalSpacing = 24;
    const cols = Math.max((totalWidth / (width + horizontalSpacing)) | 0, 1);
    related = related.slice(0, rows * cols);

    return (
      <div
        ref={this.ref}
        style={{
          fontFamily: "FaktPro-Light, sans-serif"
        }}
      >
        {related.length === 0 ? (
          undefined
        ) : (
          <h2
            style={{
              fontWeight: "400",
              fontFamily: "FaktPro-Light, sans-serif",
              fontSize: 40
            }}
          >
            {Drupal.t("Inspiration")}
          </h2>
        )}
        <div
          style={{
            display: "flex",
            justifyContent: "flex-start",
            flexWrap: "wrap"
          }}
        >
          {related.map(o => {
            const { hasPart, image, name, creator } = o;
            const { datePublished, url } = hasPart[0];
            const { pid } = hasPart[0];

            return (
              <a
                href={`/ting/object/${pid}?u_navigatedby=bibspire`}
                key={pid}
                style={{
                  textDecoration: "none",
                  display: "flex",
                  flexDirection: "column",
                  color: "black",
                  padding: `0 ${horizontalSpacing}px ${verticalSpacing}px 0`
                }}
              >
                <div
                  style={{
                    display: "flex",
                    width: width,
                    height: coverHeight,
                    backgroundImage: `url(${image})`,
                    backgroundSize: "cover",
                    margin: `0 0 ${coverTextSpacing}px 0`
                  }}
                />
                <div
                  style={{
                    display: "inline-block",
                    fontSize: fontSize + "px",
                    lineHeight: lineHeight + "px",
                    width: width
                  }}
                >
                  {name}
                  <br />
                  <strong>
                    {(creator[0] || "").replace(/\(.*\)/, "")} ({datePublished})
                  </strong>
                </div>
              </a>
            );
          })}
        </div>
      </div>
    );
  }
}

export default RecommenderWidget;


js/src/RecommendationCols.js
/*! DDB-CMS recommender widget, copyright Veduz ApS */
import React from "react";
import ReactDOM from "react-dom";

class RecommendationCols extends React.Component {
  constructor(props) {
    super(props);
    this.ref = React.createRef();
    this.state = {};
  }

  componentDidMount() {
    if (this.ref.current) {
      this.setState({ totalWidth: this.ref.current.offsetWidth });
    }
  }

  render() {
    setTimeout(() => {
      if (this.ref.current) {
        this.setState({ totalWidth: this.ref.current.offsetWidth });
      }
    }, 0);
    let { related, config } = this.props;
    if (!related || related.length === 0) {
      return null;
    }
    const totalWidth = this.state.totalWidth || 100;
    const rows = 3;
    let horizontalSpacing = 24;
    let width = +config.coverWidth || 158;
    const minCols = 4;
    const minWidth = minCols * (width + horizontalSpacing);

    if (minWidth > totalWidth) {
      horizontalSpacing *= totalWidth / minWidth;
      width *= totalWidth / minWidth;
    }
    const coverRatio = 1.398734;
    const coverHeight = width * coverRatio;
    const fontSize = 14;
    const lineHeight = 20;
    const coverTextSpacing = 8;
    const verticalSpacing = 24;
    const cols = Math.max((totalWidth / (width + horizontalSpacing)) | 0, 1);
    related = related.slice(0, rows * cols);

    return (
      <div
        ref={this.ref}
        style={{
          fontFamily: "FaktPro-Light, sans-serif"
        }}
      >
        {related.length === 0 ? (
          undefined
        ) : (
          <h2
            style={{
              fontWeight: "400",
              fontFamily: "FaktPro-Light, sans-serif",
              fontSize: 40
            }}
          >
            {Drupal.t("Inspiration")}
          </h2>
        )}
        <div
          style={{
            display: "flex",
            justifyContent: "flex-start",
            flexWrap: "wrap"
          }}
        >
          {related.map(o => {
            const { hasPart, image, name, creator } = o;
            const { datePublished, url } = hasPart[0];
            const { pid } = hasPart[0];

            return (
              <a
                href={`/ting/object/${pid}?u_navigatedby=bibspire`}
                key={pid}
                style={{
                  textDecoration: "none",
                  display: "flex",
                  flexDirection: "column",
                  color: "black",
                  padding: `0 ${horizontalSpacing}px ${verticalSpacing}px 0`
                }}
              >
                <div
                  style={{
                    display: "flex",
                    width: width,
                    height: coverHeight,
                    backgroundImage: `url(${image})`,
                    backgroundSize: "cover",
                    margin: `0 0 ${coverTextSpacing}px 0`
                  }}
                />
                <div
                  style={{
                    display: "inline-block",
                    fontSize: fontSize + "px",
                    lineHeight: lineHeight + "px",
                    width: width
                  }}
                >
                  <div
                    style={{
                      lineHeight: 16 + "px",
                      maxHeight: 32,
                      overflow: "hidden",
                      marginBottom: 4
                    }}
                  >
                    {name}
                  </div>
                  <div
                    style={{
                      lineHeight: 16 + "px",
                      maxHeight: 32,
                      overflow: "hidden"
                    }}
                  >
                    <strong>
                      {(creator[0] || "").replace(/\(.*\)/, "")} (
                      {datePublished})
                    </strong>
                  </div>
                </div>
              </a>
            );
          })}
        </div>
      </div>
    );
  }
}

export default RecommendationCols;


js/src/RecommendationLines.js
/*! DDB-CMS recommender widget, copyright Veduz ApS */
import React from "react";
import ReactDOM from "react-dom";

class RecommendationLines extends React.Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  componentDidMount() {}

  render() {
    let { related, config } = this.props;
    if (!related || related.length === 0) {
      return null;
    }
    related = related.slice(0, 10);

    return (
      <div>
        <h2 style={{ marginTop: 72 }} className="sub-menu-title">
          {Drupal.t("Prøv også:")}
        </h2>
        {related.slice(0, 10).map(({ name, image, creator, hasPart, wid }) => (
          <a
            key={wid}
            style={{ marginBottom: 20, display: "block" }}
            href={
              "/ting/object/" +
              hasPart[0].pid +
              "?u_navigatedby=bibspire_userstatus"
            }
          >
            <img
              style={{
                verticalAlign: "top",
                display: "inline",
                opacity: 0.7,
                marginRight: 10
              }}
              src={image}
              width={80}
            />
            <span
              style={{
                verticalAlign: "top",
                display: "inline-block",
                width: 180
              }}
            >
              <h3
                className="item-title has-message"
                style={{ margin: 0, padding: 0 }}
              >
                {name}
              </h3>
              <div className="item-creators">{creator && creator[0]}</div>
            </span>
          </a>
        ))}
      </div>
    );
  }
}

export default RecommendationLines;


js/src/RecommendationRows.js
/*! DDB-CMS recommender widget, copyright Veduz ApS */
import React from "react";
import ReactDOM from "react-dom";
import Mustache from "mustache";

class RecommendationRows extends React.Component {
  constructor(props) {
    super(props);
    this.ref = React.createRef();
    this.state = {};
  }

  componentDidMount() {
    if (this.ref.current) {
      this.setState({ totalWidth: this.ref.current.offsetWidth });
    }
  }

  render() {
    setTimeout(() => {
      if (this.ref.current) {
        this.setState({ totalWidth: this.ref.current.offsetWidth });
      }
    }, 0);
    let { related, config } = this.props;
    /*
    if (config.widgetType === "custom") {
      try {
        return (
          <div
            dangerouslySetInnerHTML={{
              __html: Mustache.render(config.customHtml, {
                related: related
                  .slice(0, config.customLimit || 6)
                  .map(({ hasPart, image, name, creator, datePublished }) => ({
                    title: name,
                    image,
                    link: `/ting/object/${hasPart[0].pid}?u_navigatedBy=bibspire`,
                    creator: creator,
                    year: hasPart[0].datePublished
                  }))
              })
            }}
          />
        );
      } catch (e) {
        return "Fejl i konfiguration af egen tilpasning af anbefalinger:" + e;
      }
    }
    */
    if (!related || related.length === 0) {
      return null;
    }
    const totalWidth = this.state.totalWidth || 100;
    const rows = 3;
    const horizontalSpacing = 24;
    const width = +config.coverWidth || 158;
    const coverRatio = 1.398734;
    const coverHeight = width * coverRatio;
    const fontSize = 14;
    const lineHeight = 20;
    const coverTextSpacing = 8;
    const verticalSpacing = 24;
    const cols = Math.max((totalWidth / (width + horizontalSpacing)) | 0, 1);
    related = related.slice(0, rows * cols);

    return (
      <div
        ref={this.ref}
        style={{
          fontFamily: "FaktPro-Light, sans-serif"
        }}
      >
        {related.length === 0 ? (
          undefined
        ) : (
          <h2
            style={{
              fontWeight: "400",
              fontFamily: "FaktPro-Light, sans-serif",
              fontSize: 40
            }}
          >
            {Drupal.t("Inspiration")}
          </h2>
        )}
        <div
          style={{
            display: "flex",
            justifyContent: "flex-start",
            flexWrap: "wrap"
          }}
        >
          {related.map(o => {
            const { hasPart, image, name, creator } = o;
            const { datePublished, url } = hasPart[0];
            const { pid } = hasPart[0];

            return (
              <a
                href={`/ting/object/${pid}?u_navigatedby=bibspire`}
                key={pid}
                style={{
                  textDecoration: "none",
                  display: "flex",
                  flexDirection: "column",
                  color: "black",
                  padding: `0 ${horizontalSpacing}px ${verticalSpacing}px 0`
                }}
              >
                <div
                  style={{
                    display: "flex",
                    width: width,
                    height: coverHeight,
                    backgroundImage: `url(${image})`,
                    backgroundSize: "cover",
                    margin: `0 0 ${coverTextSpacing}px 0`
                  }}
                />
                <div
                  style={{
                    display: "inline-block",
                    fontSize: fontSize + "px",
                    lineHeight: lineHeight + "px",
                    width: width
                  }}
                >
                  {name}
                  <br />
                  <strong>
                    {(creator[0] || "").replace(/\(.*\)/, "")} ({datePublished})
                  </strong>
                </div>
              </a>
            );
          })}
        </div>
      </div>
    );
  }
}

export default RecommendationRows;
